File: day05/print_queue.py
import sys

def main():
    part1 = 0
    part2 = 0
    in_rules = True
    rules = set()

    with open(sys.argv[1]) as f:
        for line in f:
            line = line.rstrip()
            if line == '':
                in_rules = False
            elif in_rules:
                before, after = [int(x) for x in line.split('|')]
                rules.add((before, after))
            else: # in reports, so check it
                pages = [int(x) for x in line.split(',')]
                ok = True
                for i in range(len(pages)-1):
                    for j in range(i+1, len(pages)):
                        if (pages[j], pages[i]) in rules:
                            ok = False
                            pages[i], pages[j] = pages[j], pages[i]

                if ok:
                    part1 += pages[len(pages)//2]
                else:
                    part2 += pages[len(pages)//2]

    print('Part 1:', part1)
    print('Part 2:', part2)

main()

File: day05/print_queue.pl
#!/usr/bin/env perl
use v5.40;

my $part1 = 0;
my $part2 = 0;
my $in_rules = 1;
my %rules;

while (<<>>) {
    chomp;
    if (!$_) {
        $in_rules = 0;
    } elsif ($in_rules) {
        $rules{$_} = 1;
    } else {
        my @pages = split ',';
        my $ok = 1;
        for my $i (0..$#pages-1) {
            for my $j ($i+1..$#pages) {
                if (defined $rules{"$pages[$j]|$pages[$i]"}) {
                    $ok = 0;
                    ($pages[$i], $pages[$j]) = ($pages[$j], $pages[$i]);
                }
            }
        }
        if ($ok) {
            $part1 += $pages[$#pages/2];
        } else {
            $part2 += $pages[$#pages/2];
        }
    }
}

say "Part 1: $part1";
say "Part 2: $part2";
